### PR TITLE
Experiment: Add the middleware `navigation/middleware` to isolate the URL change effect.

### PR DESCRIPTION
**Pull request by southp**
This PR attempts to experiment on the idea of isolating the URL change effect in a middleware, `navigation/middleware`. As a live example, I pick the piece from the password resetting flow that I am working on recently. To see it in action:

* Open an incognito window, and make sure you are not logged in.
* Visit http://calypso.localhost:3000/account-recovery.
* Randomly type anything in the text field, click on the **Get New Password** button, and you should be directed to `/account-recovery/reset-password` page.

It is currently under drafting, and there is one major question in pondering:

### In this case, URL changes also affect views.
In this case and many of other parts of our code, we are using `page.js` to [register view functions to routes](https://github.com/Automattic/wp-calypso/blob/master/client/account-recovery/index.js). Thus, changing URL actually also changes our view, disregarding our app state. I feel that our ideal state-view relationship should be a total one-on-one. That is, given a state value, we can guarantee the view. In the current draft, changing URL within the middleware doesn't reflect the fact to our state that our view is changed. Thus, there is actually no difference before and after one clicking the **Get New Password** button.

Several thoughts occur to me on this:

Maybe we should add a `route` property to our app state? Let's keep it small in this case, contained in the `accountRecovery` state tree:
```
accountRecovery: {
     route: '/account-recovery/reset-password'
}
```
If so, we could add an action, say `ACCOUNT_RECOVERY_GO_TO_RESET_PASSWORD`, populate the `route` field in the `reducer`, and maybe add one more middleware to dispatch the actual underlying URL change action:
```
export default ( { dispatch }, next, action ) => {
    if ( ACCOUNT_RECOVERY_GO_TO_RESET_PASSWORD === action.type ) {
        dispatch( navigateToRoute( '/account-recovery/reset-password' );
    } 
    return next( action );
}
```
However, we are still triggering the view change by the callback registered on routes. That is, given only the state, without any actions:
```
accountRecovery: {
     route: '/account-recovery/reset-password'
}
```

My view wouldn't be on the reset password page, because it needs to navigate to that URL to 
Am I too strict on the definition of state -> view translation which is nearly impractical? Does registering view callbacks by `page.js` play well with this scenario? Is there any essential piece I'm missing?

cc @dmsnell since I would love to have your idea on this :)

### CODE REVIEW

File: client/state/index.js [hunk 1]
```diff
@@ -111,7 +112,11 @@ export const reducer = combineReducers( {
 	wordads,
 } );
 
-const middleware = [ thunkMiddleware, noticesMiddleware ];
+const middleware = [
+	thunkMiddleware,
+	noticesMiddleware,
+	navigationMiddleware,
```
Comment (dmsnell) [inline client/state/index.js]: we definitely don't want to inject this middleware if we're not in the browser…

File: client/state/navigation/middleware.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import {
+	NAVIGATE_TO_ROUTE,
+} from 'state/action-types';
+
+export default ( store, next, action ) => {
+	if ( NAVIGATE_TO_ROUTE === action.type ) {
+		page( action.route );
+	}
+
+	return next( action );
```
Comment (dmsnell) [inline client/state/navigation/middleware.js]: I don't know enough about `page()` to review this but we want to make sure we don't mess ourselves up with the timing/sequence here

**Comment by dmsnell** [comment]
@ehg + @mcsf + @mtias you might be interested in this PR. I know there are other efforts going on with things like this.

nice work @southp

File: client/state/index.js (hunk 1 above)
Comment (southp) [inline client/state/index.js]: Fixed by 4c46e748f2228622fd3b51980b94e371c34962e9 ... _maybe_ . It is now complaining about this to me:

```
Uncaught (in promise) TypeError: Cannot redefine property: state
    at Function.defineProperty (<anonymous>)
    at eval (index.js:125)
    at eval (enhancer.js:36)
    at vendor.development.js?v=147ad937b5:43322
    at createReduxStore (index.js:307)
    at loadInitialStateFailed (initial-state.js:107)
```
It looks like [consoleDispatcher](https://github.com/Automattic/wp-calypso/blob/master/client/state/console-dispatch/index.js#L72) got called twice after this change. I have no idea why :(

@dmsnell Could you help me take a look?

**Comment by ehg** [comment]
I think isomorphic @ockham would be interested in this :)

**Comment by ockham** [comment]
> I think isomorphic @ockham would be interested in this :)

Yeah, as well as @aduth 😄 .

While we've tossed around a similar idea (mapping routes to Redux state), @ehg is correct that we'd ideally want something that works on the server side as well. OTOH, what this PR does looks like a basic Redux router, similar to [`redux-routing`](https://github.com/callum/redux-routing) which I think is what @aduth was experimenting with.

Finally, this experiment only covers explicit `page()` calls, right? `href`s are still going to be intercepted by `page` and wouldn't cause a Redux state change.

At this point, I'd be wary of introducing yet another way to route without thinking through the implications for the entirety of Calypso (including those aspects I pointed out above).

**Comment by aduth** [comment]
Yeah, I'm totally on board with Redux-based routing state and actions, but there's quite a few complexities in Calypso to overcome with making it work well with `page.js` (and even more to eliminate `page.js` altogether).

>If so, we could add an action, say `ACCOUNT_RECOVERY_GO_TO_RESET_PASSWORD`, populate the `route` field in the `reducer`, and maybe add one more middleware to dispatch the actual underlying URL change action:

I don't think this needs to happen in the middleware. This could be a separate action creator which simply wraps the underlying `navigateToRoute`:

```js
export function navigateToResetPassword() {
	return navigateToRoute( '/account-recovery/reset-password' );
}
```

>OTOH, what this PR does looks like a basic Redux router, similar to [`redux-routing`](https://github.com/callum/redux-routing) which I think is what @aduth was experimenting with.

Lately I've just been rolling my own. A few links for reference in case it's useful:

- https://github.com/aduth/dones/tree/master/src/state/routing
- https://github.com/aduth/dones/blob/master/src/state/selectors/get-matched-route.js
- https://github.com/aduth/dones/blob/master/src/routes/index.js
- https://github.com/aduth/dones/blob/master/src/components/link/index.js

>At this point, I'd be wary of introducing yet another way to route without thinking through the implications for the entirety of Calypso (including those aspects I pointed out above).

Another consideration is preserving or accounting for the order in which page.js middlewares are called vs. Redux middlewares. Right now our `setRouteAction` dispatch occurs somewhat awkwardly in the middle of the page.js middleware stack:

https://github.com/Automattic/wp-calypso/blob/fda96ee/client/boot/index.js#L363-L369

**Comment by dmsnell** [comment]
@southp in any case, @blowery has done some work in #10897 that would affect any additional middleware injections.

thanks all for jumping in on this. I anticipated that there would be non-obvious considerations regarding `page.js`

**Comment by dmsnell** [comment]
@southp and @aduth and @ockham should we close this PR and table the discussion for a much larger effort (in the future) to move towards state-based navigation?

**Comment by ockham** [comment]
> @southp and @aduth and @ockham should we close this PR and table the discussion for a much larger effort (in the future) to move towards state-based navigation?

Yeah, sounds good.

**Comment by southp** [comment]
Totally :)

**Pull request closed without merging**